### ISSUE TITLE
Manage nodes on different networks with no intra node connectivity

### ISSUE THREAD

**Issue by dylanenabled** [rancher/k3s#261]
Thanks for this great project! It seems like it could be very useful and I'm evaluating it for future use. I have an edge computing use case where each node (edge computing device) is installed on separate networks (for different customers of our compute module) and I wish to manage the pods/containers running on them via kubernetes. 

The nodes have no direct connectivity to each other (each node is behind a NAT and cannot have ports opened to the internet for security reasons) and can only access the master node via outbound connections to the internet. 

I note in your documentation that flannel requires nodes to be able to communicate with each other via UDP - is this required for the network to function? If so it seems to limit the potential of this for managing remote edge computing devices. I do not have the option of opening ports on each network for the edge nodes to communicate directly, but there is no requirement for nodes to communicate with each other (each node is collecting data from IoT devices via bluetooth then processing and passing that to other services in the cloud).

So in summary:

1. Is node-to-node communication required for k3s to function?
2. If so, I'd like to request a feature to remove the requirement (via configuration) for certain nodes.
[](url)

**Comment by aaliddell** [comment]
Without node->node links, both k8s and k3s will 'run', but pod networking will not; this is a fundamental constraint of the upstream Kubernetes network model that all drivers are built to: https://kubernetes.io/docs/concepts/cluster-administration/networking/#the-kubernetes-network-model

In general, Kubernetes is designed such that each node is essentially fungible, and inter-node communication is relatively cheap. i.e a pod can generally be scheduled on any node and just work without complex placement constraints (although you can add constraints for performance reasons etc, they shouldn't really be used for working around a segmented network). What this means is that clusters typically have closely colocated nodes with a full interconnected mesh of inter-node links.

In theory, you could get around this by having each node join a centralised VPN with your master, then running the overlay network on top of that, but performance and stability will suffer and VXLAN mode for Flannel likely won't work.

**Comment by erikwilson** [comment]
Another way to consider organizing might be to have each node as a single-node cluster (or if a customer has many nodes then great), and create a multi-cluster app around that model. It would probably work well for something like map/reduce, but maybe not great for other computational models that benefit from fast network interconnects.

**Comment by erikwilson** [comment]
Also, you may want to look into how we handle imported clusters in Rancher, normally with provisioning we SSH to a machine to provision but with imported clusters (which will work with k3s) we run a kubectl commands on the cluster to communicate back to the server.

I am curious about your edge compute scenario also, the classic edge example might use something like mqtt and lamda functions to process locally and push upstream, but it sounds like you are pushing computation down to the nodes to solve large problems.

**Comment by erikwilson** [comment]
Sorry - I really need to work on my reading comprehension. :) For a scenario like `(each node is collecting data from IoT devices via bluetooth then processing and passing that to other services in the cloud)` it is probably easiest to consider it as a single node cluster gateway device, or as a multi-node cluster if there are multiple devices covering the same area for the same customer. Rancher was designed to help with the management aspect, and we are planning on adding additional features to assist with those types of deployments.

**Comment by dylanenabled** [comment]
Great! Thanks for the responses. I'll look into Rancher some more and see if that suits. I haven't used k8s at all (though we use containers heavily), so any links you can give on setting up node clusters would certainly help!
